File: static/js/experiment_config.js
/**
 * Experiment configuration preview.
 * Power / exact-vs-sampled math matches utils/permutation_test.py (power_guardrail).
 * Copy strings come from window.FOCALPROMPT_COPY when present.
 */
(function (global) {
    'use strict';

    var DEFAULT_N_PERMUTATIONS = 10000;
    var DEFAULT_ALPHA = 0.05;
    var DEFAULT_TEMPERATURE = 0.7;
    var DEFAULT_N_BASELINE = 10;
    var DEFAULT_N_ABLATED = 5;
    var N_BASELINE_MIN = 5;
    var N_BASELINE_MAX = 50;
    var N_ABLATED_MIN = 3;
    var N_ABLATED_MAX = 25;

    function copy() {
        return global.FOCALPROMPT_COPY || {};
    }

    function combinations(n, k) {
        if (k < 0 || k > n) return 0;
        k = Math.min(k, n - k);
        var r = 1;
        for (var i = 1; i <= k; i++) {
            r = r * (n - k + i) / i;
        }
        return Math.round(r);
    }

    function nLabelAssignments(nBaseline, nAblated) {
        nBaseline = Number(nBaseline);
        nAblated = Number(nAblated);
        var n = nBaseline + nAblated;
        if (nBaseline < 0 || nAblated < 0 || n === 0) return 0;
        return combinations(n, nAblated);
    }

    function minAchievablePvalue(nBaseline, nAblated, nPermutations) {
        nPermutations = nPermutations == null ? DEFAULT_N_PERMUTATIONS : nPermutations;
        var nExact = nLabelAssignments(nBaseline, nAblated);
        if (nExact === 0) return 1;
        if (nExact <= nPermutations) return 1 / nExact;
        return 1 / (1 + nPermutations);
    }

    function usesExactEnumeration(nBaseline, nAblated, nPermutations) {
        nPermutations = nPermutations == null ? DEFAULT_N_PERMUTATIONS : nPermutations;
        var nExact = nLabelAssignments(nBaseline, nAblated);
        return nExact > 0 && nExact <= nPermutations;
    }

    function testTypeForDesign(nBaseline, nAblated, nPermutations) {
        return usesExactEnumeration(nBaseline, nAblated, nPermutations) ? 'exact' : 'sampled';
    }

    function monteCarloPvalueSe(p, nPermutations) {
        p = p == null ? 0.05 : p;
        nPermutations = nPermutations == null ? DEFAULT_N_PERMUTATIONS : nPermutations;
        if (nPermutations <= 0) return 1;
        return Math.sqrt(p * (1 - p) / nPermutations);
    }

    function powerGuardrail(nBaseline, nAblated, nFoci, alpha, nPermutations) {
        alpha = alpha == null ? DEFAULT_ALPHA : alpha;
        nPermutations = nPermutations == null ? DEFAULT_N_PERMUTATIONS : nPermutations;
        nBaseline = Number(nBaseline);
        nAblated = Number(nAblated);
        nFoci = Number(nFoci);
        var minP = minAchievablePvalue(nBaseline, nAblated, nPermutations);
        var nAssignments = nLabelAssignments(nBaseline, nAblated);
        var exact = usesExactEnumeration(nBaseline, nAblated, nPermutations);
        var testType = exact ? 'exact' : 'sampled';
        if (nFoci <= 0) {
            return {
                min_p: minP,
                threshold: null,
                can_reach_significance: null,
                n_foci: nFoci,
                exact: exact,
                n_assignments: nAssignments,
                test_type: testType
            };
        }
        var threshold = alpha / nFoci;
        return {
            min_p: minP,
            threshold: threshold,
            can_reach_significance: minP <= threshold,
            n_foci: nFoci,
            exact: exact,
            n_assignments: nAssignments,
            test_type: testType
        };
    }

    function stochasticTemperatureMessage(temperature) {
        var C = copy();
        if (C.STOCHASTIC_TEMPERATURE_TEMPLATE) {
            return C.STOCHASTIC_TEMPERATURE_TEMPLATE.replace('{temperature}', String(temperature));
        }
        return (
            'Permutation test requires output stochasticity: temperature must be > 0 (got ' +
            temperature +
            '). Repeated samples of the same prompt must be allowed to vary; set temperature above 0.'
        );
    }

    function temperatureRejection(temperature) {
        if (temperature === null || temperature === undefined || temperature === '') {
            return stochasticTemperatureMessage(temperature);
        }
        var t = Number(temperature);
        if (!Number.isFinite(t) || t <= 0) {
            return stochasticTemperatureMessage(Number.isFinite(t) ? t : temperature);
        }
        return null;
    }

    function suggestedSampleSizes(temperature) {
        var t = Number(temperature);
        if (t > 1.0) return { n_baseline: 15, n_ablated: 8 };
        return { n_baseline: 10, n_ablated: 5 };
    }

    function clampNBaseline(value) {
        var n = parseInt(value, 10);
        if (!Number.isFinite(n)) n = DEFAULT_N_BASELINE;
        return Math.max(N_BASELINE_MIN, Math.min(N_BASELINE_MAX, n));
    }

    function clampNAblated(value) {
        var n = parseInt(value, 10);
        if (!Number.isFinite(n)) n = DEFAULT_N_ABLATED;
        return Math.max(N_ABLATED_MIN, Math.min(N_ABLATED_MAX, n));
    }

    function modelCallCount(nBaseline, nAblated, nAttributableFoci) {
        return Number(nBaseline) + Number(nAblated) * Number(nAttributableFoci);
    }

    function countPreviewAttributable(foci) {
        if (!foci || !foci.length) return 0;
        var n = 0;
        for (var i = 0; i < foci.length; i++) {
            if (!foci[i].is_dynamic) n += 1;
        }
        return n;
    }

    function formatCostLine(nBaseline, nAblated, nAttributableFoci, fociTagged) {
        var C = copy();
        if (!fociTagged) {
            return (C.COST_LINE_FORMULA || 'This experiment will make {n_baseline} + {n_ablated} × n_foci model calls.')
                .replace('{n_baseline}', String(nBaseline))
                .replace('{n_ablated}', String(nAblated));
        }
        var nCalls = modelCallCount(nBaseline, nAblated, nAttributableFoci || 0);
        return (C.COST_LINE_COUNTED || 'This experiment will make {n_calls} model calls.')
            .replace('{n_calls}', String(nCalls));
    }

    function formatPermutationDisclosure(nBaseline, nAblated, nPermutations) {
        var C = copy();
        nPermutations = nPermutations == null ? DEFAULT_N_PERMUTATIONS : nPermutations;
        if (usesExactEnumeration(nBaseline, nAblated, nPermutations)) {
            var nAssign = nLabelAssignments(nBaseline, nAblated);
            return (C.EXACT_DISCLOSURE_TEMPLATE ||
                'Significance: exact test ({n_assignments} enumerated group assignments)')
                .replace('{n_assignments}', nAssign.toLocaleString('en-US'));
        }
        var se = monteCarloPvalueSe(0.05, nPermutations);
        return (C.SAMPLED_DISCLOSURE_TEMPLATE ||
            'Significance: 10,000 sampled permutations (p-value margin ~±{se})')
            .replace('{se}', se.toFixed(3));
    }

    function formatPowerPreviewLine(nBaseline, nAblated, nFoci, alpha, nPermutations) {
        var C = copy();
        var info = powerGuardrail(nBaseline, nAblated, nFoci, alpha, nPermutations);
        if (info.can_reach_significance === null) return null;
        if (info.can_reach_significance) {
            return C.POWER_OK || 'This design can detect effects at your significance level';
        }
        return (C.POWER_FAIL ||
            'With {n_foci} foci, this design cannot reach significance after correction. Increase samples.')
            .replace('{n_foci}', String(nFoci));
    }

    function formatTemperature(t) {
        return Number(t).toFixed(1);
    }

    function formatAblationLoading(temperature, nBaseline, nAblated, nFoci) {
        var C = copy();
        nFoci = Number(nFoci) || 0;
        var fociWord = nFoci === 1 ? 'focus' : 'foci';
        return (C.ABLATION_LOADING_TEMPLATE ||
            'Running ablation analysis at temperature {temperature}: {n_baseline} baseline samples and {n_ablated} ablated samples for each of {n_foci} {foci_word} ({n_calls} model calls). This may take several minutes.')
            .replace('{temperature}', formatTemperature(temperature))
            .replace('{n_baseline}', String(nBaseline))
            .replace('{n_ablated}', String(nAblated))
            .replace('{n_foci}', String(nFoci))
            .replace('{foci_word}', fociWord)
            .replace('{n_calls}', String(modelCallCount(nBaseline, nAblated, nFoci)));
    }

    function formatBatchLoading(nPairs, temperature, nBaseline, nAblated) {
        var C = copy();
        nPairs = Number(nPairs) || 0;
        var pairsWord = nPairs === 1 ? 'pair' : 'pairs';
        return (C.BATCH_LOADING_TEMPLATE ||
            'Running batch analysis on {n_pairs} {pairs_word} at temperature {temperature}: {n_baseline} baseline samples and {n_ablated} ablated samples per focus per pair. This may take a long time.')
            .replace('{n_pairs}', String(nPairs))
            .replace('{pairs_word}', pairsWord)
            .replace('{temperature}', formatTemperature(temperature))
            .replace('{n_baseline}', String(nBaseline))
            .replace('{n_ablated}', String(nAblated));
    }

    function formatRunHeader(temperature, nBaseline, nAblated, testType) {
        var C = copy();
        var kind = testType === 'exact' ? 'exact' : 'sampled';
        return (C.RUN_HEADER_TEMPLATE ||
            'Run at temperature {t}, {n_baseline}+{n_ablated} samples per focus, {test_type} test.')
            .replace('{t}', formatTemperature(temperature))
            .replace('{n_baseline}', String(nBaseline))
            .replace('{n_ablated}', String(nAblated))
            .replace('{test_type}', kind);
    }

    function formatRunHeaderFromData(data) {
        data = data || {};
        var nBaseline = Number(data.n_baseline || data.num_baseline_samples || 10);
        var nAblated = Number(data.n_ablated || 5);
        var nPerm = Number(data.n_permutations || DEFAULT_N_PERMUTATIONS);
        var temperature = data.temperature != null ? data.temperature : DEFAULT_TEMPERATURE;
        var testType = data.test_type || testTypeForDesign(nBaseline, nAblated, nPerm);
        return formatRunHeader(temperature, nBaseline, nAblated, testType);
    }

    function readStateFromRoot(root) {
        var tempEl = root.querySelector('.exp-temperature');
        var baseEl = root.querySelector('.exp-n-baseline');
        var ablEl = root.querySelector('.exp-n-ablated');
        return {
            temperature: tempEl ? Number(tempEl.value) : DEFAULT_TEMPERATURE,
            n_baseline: baseEl ? clampNBaseline(baseEl.value) : DEFAULT_N_BASELINE,
            n_ablated: ablEl ? clampNAblated(ablEl.value) : DEFAULT_N_ABLATED
        };
    }

    var sharedState = {
        temperature: DEFAULT_TEMPERATURE,
        n_baseline: DEFAULT_N_BASELINE,
        n_ablated: DEFAULT_N_ABLATED
    };

    function fociForRoot(root) {
        var source = root.getAttribute('data-foci-source');
        if (source === 'batch') return global.batchFoci || [];
        return global.foci || [];
    }

    function refreshRoot(root) {
        var C = copy();
        var state = sharedState;
        var tempEl = root.querySelector('.exp-temperature');
        var baseEl = root.querySelector('.exp-n-baseline');
        var ablEl = root.querySelector('.exp-n-ablated');
        if (tempEl && document.activeElement !== tempEl) tempEl.value = formatTemperature(state.temperature);
        if (baseEl && document.activeElement !== baseEl) baseEl.value = String(state.n_baseline);
        if (ablEl && document.activeElement !== ablEl) ablEl.value = String(state.n_ablated);

        var help = root.querySelector('.exp-temperature-help');
        if (help) help.textContent = C.TEMPERATURE_HELP || '';
        var high = root.querySelector('.exp-temperature-high');
        if (high) {
            high.textContent = C.TEMPERATURE_HIGH || '';
            if (Number(state.temperature) >= 1.0) high.classList.remove('hidden');
            else high.classList.add('hidden');
        }
        var tempErr = root.querySelector('.exp-temperature-error');
        var rejection = temperatureRejection(state.temperature);
        if (tempErr) {
            if (rejection) {
                tempErr.textContent = rejection;
                tempErr.classList.remove('hidden');
            } else {
                tempErr.textContent = '';
                tempErr.classList.add('hidden');
            }
        }

        var foci = fociForRoot(root);
        var fociTagged = !!(foci && foci.length);
        var nAttr = countPreviewAttributable(foci);
        var costEl = root.querySelector('.exp-cost-line');
        if (costEl) {
            costEl.textContent = formatCostLine(state.n_baseline, state.n_ablated, nAttr, fociTagged);
        }

        var sug = suggestedSampleSizes(state.temperature);
        var chip = root.querySelector('.exp-suggestion-chip');
        if (chip) {
            chip.textContent = (C.SUGGESTION_LABEL || 'suggested for this temperature') +
                ' (' + sug.n_baseline + ' / ' + sug.n_ablated + ')';
            chip.setAttribute('title', C.SUGGESTION_TOOLTIP || '');
            chip.setAttribute('aria-label', (C.SUGGESTION_TOOLTIP || '') + ' Apply ' + sug.n_baseline + ' baseline and ' + sug.n_ablated + ' ablated.');
            chip.dataset.nBaseline = String(sug.n_baseline);
            chip.dataset.nAblated = String(sug.n_ablated);
        }
        var tip = root.querySelector('.exp-suggestion-tooltip');
        if (tip) tip.textContent = C.SUGGESTION_TOOLTIP || '';

        var powerEl = root.querySelector('.exp-power-line');
        if (powerEl) {
            var line = formatPowerPreviewLine(state.n_baseline, state.n_ablated, nAttr);
            if (!line) {
                powerEl.textContent = '';
                powerEl.className = 'exp-power-line hidden';
            } else {
                powerEl.textContent = line;
                powerEl.className = 'exp-power-line ' +
                    (line === (C.POWER_OK || 'This design can detect effects at your significance level')
                        ? 'exp-power-ok'
                        : 'exp-power-fail');
            }
        }

        var disc = root.querySelector('.exp-permutation-disclosure');
        if (disc) {
            disc.textContent = formatPermutationDisclosure(state.n_baseline, state.n_ablated);
        }
    }

    function refreshAll() {
        var roots = document.querySelectorAll('.experiment-config');
        for (var i = 0; i < roots.length; i++) refreshRoot(roots[i]);
    }

    function applyState(partial) {
        if (partial.temperature != null) sharedState.temperature = Number(partial.temperature);
        if (partial.n_baseline != null) sharedState.n_baseline = clampNBaseline(partial.n_baseline);
        if (partial.n_ablated != null) sharedState.n_ablated = clampNAblated(partial.n_ablated);
        refreshAll();
    }

    function getState() {
        return {
            temperature: sharedState.temperature,
            n_baseline: sharedState.n_baseline,
            n_ablated: sharedState.n_ablated,
            n_permutations: DEFAULT_N_PERMUTATIONS,
            test_type: testTypeForDesign(sharedState.n_baseline, sharedState.n_ablated)
        };
    }

    function bind() {
        if (typeof document === 'undefined') return;
        var roots = document.querySelectorAll('.experiment-config');
        for (var i = 0; i < roots.length; i++) {
            (function (root) {
                root.addEventListener('input', function (e) {
                    var t = e.target;
                    if (t.classList.contains('exp-temperature')) {
                        sharedState.temperature = Number(t.value);
                    } else if (t.classList.contains('exp-n-baseline')) {
                        sharedState.n_baseline = clampNBaseline(t.value);
                    } else if (t.classList.contains('exp-n-ablated')) {
                        sharedState.n_ablated = clampNAblated(t.value);
                    }
                    refreshAll();
                });
                root.addEventListener('change', function (e) {
                    var t = e.target;
                    if (t.classList.contains('exp-n-baseline')) {
                        t.value = String(clampNBaseline(t.value));
                        sharedState.n_baseline = clampNBaseline(t.value);
                    }
                    if (t.classList.contains('exp-n-ablated')) {
                        t.value = String(clampNAblated(t.value));
                        sharedState.n_ablated = clampNAblated(t.value);
                    }
                    refreshAll();
                });
                var chip = root.querySelector('.exp-suggestion-chip');
                if (chip && !chip._bound) {
                    chip._bound = true;
                    chip.addEventListener('click', function () {
                        var sug = suggestedSampleSizes(sharedState.temperature);
                        sharedState.n_baseline = sug.n_baseline;
                        sharedState.n_ablated = sug.n_ablated;
                        refreshAll();
                    });
                }
            })(roots[i]);
        }
        refreshAll();
    }

    var api = {
        DEFAULT_N_PERMUTATIONS: DEFAULT_N_PERMUTATIONS,
        nLabelAssignments: nLabelAssignments,
        minAchievablePvalue: minAchievablePvalue,
        usesExactEnumeration: usesExactEnumeration,
        testTypeForDesign: testTypeForDesign,
        monteCarloPvalueSe: monteCarloPvalueSe,
        powerGuardrail: powerGuardrail,
        temperatureRejection: temperatureRejection,
        stochasticTemperatureMessage: stochasticTemperatureMessage,
        suggestedSampleSizes: suggestedSampleSizes,
        clampNBaseline: clampNBaseline,
        clampNAblated: clampNAblated,
        modelCallCount: modelCallCount,
        countPreviewAttributable: countPreviewAttributable,
        formatCostLine: formatCostLine,
        formatPermutationDisclosure: formatPermutationDisclosure,
        formatPowerPreviewLine: formatPowerPreviewLine,
        formatAblationLoading: formatAblationLoading,
        formatBatchLoading: formatBatchLoading,
        formatRunHeader: formatRunHeader,
        formatRunHeaderFromData: formatRunHeaderFromData,
        getState: getState,
        applyState: applyState,
        refreshAll: refreshAll,
        bind: bind
    };

    global.FocalPromptExperiment = api;
    if (typeof module !== 'undefined' && module.exports) {
        module.exports = api;
    }
})(typeof window !== 'undefined' ? window : global);

File: utils/experiment_config.py
#!/usr/bin/env python3
"""
Experiment configuration: copy, cost, sample-size suggestions, and live preview.

Preview math calls the same `power_guardrail` / permutation-regime helpers as
the API. No change to the permutation test itself.
"""

from __future__ import annotations

from math import comb
from typing import Any, Dict, Mapping, Optional, Sequence, Tuple

from utils.permutation_test import (
    DEFAULT_ALPHA,
    DEFAULT_N_PERMUTATIONS,
    min_achievable_pvalue,
    monte_carlo_pvalue_se,
    power_guardrail,
    stochastic_temperature_message,
    design_test_type,
    uses_exact_enumeration,
)

STOCHASTIC_TEMPERATURE_TEMPLATE = (
    "Permutation test requires output stochasticity: temperature must be > 0 "
    "(got {temperature}). Repeated samples of the same prompt must be allowed "
    "to vary; set temperature above 0."
)

DEFAULT_TEMPERATURE = 0.7
TEMPERATURE_MIN = 0.1
TEMPERATURE_MAX = 2.0
TEMPERATURE_STEP = 0.1
DEFAULT_N_BASELINE = 10
DEFAULT_N_ABLATED = 5
N_BASELINE_MIN = 5
N_BASELINE_MAX = 50
N_ABLATED_MIN = 3
N_ABLATED_MAX = 25

TEMPERATURE_HELP = (
    "Use the temperature your prompt runs at in production. Results describe "
    "the model's behaviour at this temperature only."
)

TEMPERATURE_HIGH = (
    "High temperature widens normal output variation, so subtle effects need "
    "more samples to detect."
)

SUGGESTION_LABEL = "suggested for this temperature"

SUGGESTION_TOOLTIP = (
    "A heuristic starting point. If results warn about power or sit near the "
    "threshold, increase ablated samples."
)

COST_LINE_COUNTED = "This experiment will make {n_calls} model calls."

COST_LINE_FORMULA = (
    "This experiment will make {n_baseline} + {n_ablated} × n_foci model calls."
)

POWER_OK = "This design can detect effects at your significance level"

POWER_FAIL = (
    "With {n_foci} foci, this design cannot reach significance after correction. "
    "Increase samples."
)

EXACT_DISCLOSURE_TEMPLATE = (
    "Significance: exact test ({n_assignments} enumerated group assignments)"
)

SAMPLED_DISCLOSURE_TEMPLATE = (
    "Significance: 10,000 sampled permutations (p-value margin ~±{se})"
)

RUN_HEADER_TEMPLATE = (
    "Run at temperature {t}, {n_baseline}+{n_ablated} samples per focus, "
    "{test_type} test."
)

EXPERIMENT_COPY = {
    'TEMPERATURE_HELP': TEMPERATURE_HELP,
    'TEMPERATURE_HIGH': TEMPERATURE_HIGH,
    'SUGGESTION_LABEL': SUGGESTION_LABEL,
    'SUGGESTION_TOOLTIP': SUGGESTION_TOOLTIP,
    'COST_LINE_COUNTED': COST_LINE_COUNTED,
    'COST_LINE_FORMULA': COST_LINE_FORMULA,
    'POWER_OK': POWER_OK,
    'POWER_FAIL': POWER_FAIL,
    'EXACT_DISCLOSURE_TEMPLATE': EXACT_DISCLOSURE_TEMPLATE,
    'SAMPLED_DISCLOSURE_TEMPLATE': SAMPLED_DISCLOSURE_TEMPLATE,
    'RUN_HEADER_TEMPLATE': RUN_HEADER_TEMPLATE,
    'STOCHASTIC_TEMPERATURE_TEMPLATE': STOCHASTIC_TEMPERATURE_TEMPLATE,
}


def format_temperature(temperature: float) -> str:
    return f'{float(temperature):.1f}'


def temperature_rejection(temperature: Any) -> Optional[str]:
    """Client- and server-side: None if ok, else the stochasticity explanation."""
    if temperature is None:
        return stochastic_temperature_message(temperature)
    try:
        t = float(temperature)
    except (TypeError, ValueError):
        return stochastic_temperature_message(temperature)
    if t <= 0:
        return stochastic_temperature_message(temperature)
    return None


def suggested_sample_sizes(temperature: float) -> Tuple[int, int]:
    """Heuristic starting point. t < 0.5 and 0.5–1.0 → 10/5; t > 1.0 → 15/8."""
    t = float(temperature)
    if t > 1.0:
        return (15, 8)
    return (10, 5)


def clamp_n_baseline(value: Any) -> int:
    try:
        n = int(value)
    except (TypeError, ValueError):
        n = DEFAULT_N_BASELINE
    return max(N_BASELINE_MIN, min(N_BASELINE_MAX, n))


def clamp_n_ablated(value: Any) -> int:
    try:
        n = int(value)
    except (TypeError, ValueError):
        n = DEFAULT_N_ABLATED
    return max(N_ABLATED_MIN, min(N_ABLATED_MAX, n))


def model_call_count(n_baseline: int, n_ablated: int, n_attributable_foci: int) -> int:
    return int(n_baseline) + int(n_ablated) * int(n_attributable_foci)


def count_preview_attributable(foci: Optional[Sequence[Mapping[str, Any]]]) -> int:
    """Client preview: tagged foci that are not runtime slots."""
    if not foci:
        return 0
    n = 0
    for focus in foci:
        if not focus.get('is_dynamic'):
            n += 1
    return n


def format_cost_line(
    n_baseline: int,
    n_ablated: int,
    n_attributable_foci: Optional[int] = None,
    foci_tagged: bool = False,
) -> str:
    if not foci_tagged:
        return COST_LINE_FORMULA.format(
            n_baseline=int(n_baseline),
            n_ablated=int(n_ablated),
        )
    n_calls = model_call_count(n_baseline, n_ablated, int(n_attributable_foci or 0))
    return COST_LINE_COUNTED.format(n_calls=n_calls)


def format_permutation_disclosure(
    n_baseline: int,
    n_ablated: int,
    n_permutations: int = DEFAULT_N_PERMUTATIONS,
) -> str:
    if uses_exact_enumeration(n_baseline, n_ablated, n_permutations):
        n_assignments = comb(int(n_baseline) + int(n_ablated), int(n_ablated))
        return EXACT_DISCLOSURE_TEMPLATE.format(
            n_assignments=f'{n_assignments:,}'
        )
    se = monte_carlo_pvalue_se(0.05, n_permutations)
    return SAMPLED_DISCLOSURE_TEMPLATE.format(se=f'{se:.3f}')


def format_power_preview_line(
    n_baseline: int,
    n_ablated: int,
    n_foci: int,
    alpha: float = DEFAULT_ALPHA,
    n_permutations: int = DEFAULT_N_PERMUTATIONS,
) -> Optional[str]:
    """Green/amber line, or None when focus count is unknown."""
    info = power_guardrail(
        n_baseline, n_ablated, n_foci, alpha=alpha, n_permutations=n_permutations
    )
    if info['can_reach_significance'] is None:
        return None
    if info['can_reach_significance']:
        return POWER_OK
    return POWER_FAIL.format(n_foci=int(n_foci))


def format_run_header(
    temperature: float,
    n_baseline: int,
    n_ablated: int,
    test_type: str,
) -> str:
    kind = 'exact' if test_type == 'exact' else 'sampled'
    return RUN_HEADER_TEMPLATE.format(
        t=format_temperature(temperature),
        n_baseline=int(n_baseline),
        n_ablated=int(n_ablated),
        test_type=kind,
    )


def experiment_preview(
    temperature: float,
    n_baseline: int,
    n_ablated: int,
    foci: Optional[Sequence[Mapping[str, Any]]] = None,
    alpha: float = DEFAULT_ALPHA,
    n_permutations: int = DEFAULT_N_PERMUTATIONS,
) -> Dict[str, Any]:
    """Single payload for tests and any server-side preview."""
    foci_tagged = bool(foci)
    n_attr = count_preview_attributable(foci)
    n_foci_for_bh = n_attr
    guard = power_guardrail(
        n_baseline, n_ablated, n_foci_for_bh, alpha=alpha, n_permutations=n_permutations
    )
    suggested = suggested_sample_sizes(temperature)
    return {
        'temperature': float(temperature),
        'n_baseline': int(n_baseline),
        'n_ablated': int(n_ablated),
        'n_attributable_foci': n_attr,
        'foci_tagged': foci_tagged,
        'cost_line': format_cost_line(
            n_baseline, n_ablated, n_attr, foci_tagged=foci_tagged
        ),
        'model_calls': (
            model_call_count(n_baseline, n_ablated, n_attr) if foci_tagged else None
        ),
        'suggested_n_baseline': suggested[0],
        'suggested_n_ablated': suggested[1],
        'high_temperature': float(temperature) >= 1.0,
        'temperature_error': temperature_rejection(temperature),
        'permutation_disclosure': format_permutation_disclosure(
            n_baseline, n_ablated, n_permutations
        ),
        'test_type': design_test_type(n_baseline, n_ablated, n_permutations),
        'power_line': format_power_preview_line(
            n_baseline, n_ablated, n_foci_for_bh, alpha=alpha, n_permutations=n_permutations
        ),
        'power_ok': guard['can_reach_significance'],
        'min_p': guard['min_p'],
        'run_header': format_run_header(
            temperature,
            n_baseline,
            n_ablated,
            design_test_type(n_baseline, n_ablated, n_permutations),
        ),
        'min_achievable_p': min_achievable_pvalue(n_baseline, n_ablated, n_permutations),
    }
